### PR TITLE
[feature](struct-type) adjust the vexpr for struct type

### PR DESCRIPTION
**Pull request by carlvinhust2012**
# Proposed changes
1. this pr is used to adjust the vexpr for struct type when we import data.
2. load the json as below to the table which include the struct column.
{ "id" : 0, "s1" : { "name": "jack", "age": 20 } }]

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
5. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
6. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Type.java [hunk 1]
```diff
@@ -169,8 +169,29 @@ public abstract class Type {
         arraySubTypes.add(STRING);
 
         structSubTypes = Lists.newArrayList();
+        structSubTypes.add(TINYINT);
+        structSubTypes.add(SMALLINT);
         structSubTypes.add(INT);
+        structSubTypes.add(BIGINT);
+        structSubTypes.add(LARGEINT);
```
Comment (cambyzju) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Type.java:176]: structSubTypes.addAll(integerTypes);

Comment (cambyzju) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Type.java:176]: or addAll(numericTypes);

File: be/src/vec/data_types/data_type_struct.cpp [hunk 2]
```diff
@@ -91,6 +80,88 @@ std::string DataTypeStruct::do_get_name() const {
...
 
+Status DataTypeStruct::from_string(ReadBuffer& rb, IColumn* column) const {
+    DCHECK(!rb.eof());
+    auto* struct_column = assert_cast<ColumnStruct*>(column);
+
+    if (*rb.position() != '{') {
+        return Status::InvalidArgument("Struct does not start with '{' character, found '{}'",
+                                       *rb.position());
+    }
+    if (*(rb.end() - 1) != '}') {
```
Comment (cambyzju) [inline be/src/vec/data_types/data_type_struct.cpp:91]: size == 1 may crash?

File: be/src/vec/data_types/data_type_struct.cpp [hunk 3]
```diff
@@ -91,6 +80,88 @@ std::string DataTypeStruct::do_get_name() const {
...
+        ReadBuffer field_rb(rb.position(), field_len);
+        size_t len = 0;
+        auto start_rb = field_rb.position();
+        while (!field_rb.eof() && *start_rb != ':') {
+            len++;
+            start_rb++;
+        }
+        ReadBuffer field(field_rb.position() + len + 1, field_rb.count() - len - 1);
+
+        if (*field.position() != '"' && *field.end() != '"') {
```
Comment (cambyzju) [inline be/src/vec/data_types/data_type_struct.cpp:119]: if `field_size < 2 || *field.position() != '"' || *field.end() != '"'` ?

File: be/src/vec/data_types/data_type_factory.cpp [hunk 4]
```diff
@@ -20,6 +20,10 @@
 
 #include "vec/data_types/data_type_factory.hpp"
 
+#include "vec/data_types/data_type_hll.h"
+#include "vec/data_types/data_type_jsonb.h"
```
Comment (cambyzju) [inline be/src/vec/data_types/data_type_factory.cpp:24]: hll and jsonb already included in data_type_factory.hpp

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Type.java [hunk 5]
```diff
@@ -169,8 +169,29 @@ public abstract class Type {
...
+        structSubTypes.add(VARCHAR);
         structSubTypes.add(STRING);
+        structSubTypes.add(CHAR);
+        structSubTypes.add(DATE);
+        structSubTypes.add(DATETIME);
+        structSubTypes.add(DATEV2);
+        structSubTypes.add(DATETIMEV2);
+        structSubTypes.add(TIME);
+        structSubTypes.add(TIMEV2);
+        structSubTypes.add(JSONB);
```
Comment (cambyzju) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Type.java:190]: do not add complex type  now for STRUCT

File: be/src/vec/data_types/data_type_struct.cpp (hunk 2 above)
Comment (carlvinhust2012) [reply be/src/vec/data_types/data_type_struct.cpp:91]: done

File: be/src/vec/data_types/data_type_factory.cpp (hunk 4 above)
Comment (carlvinhust2012) [reply be/src/vec/data_types/data_type_factory.cpp:24]: done

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Type.java (hunk 5 above)
Comment (carlvinhust2012) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Type.java:190]: done

File: be/src/vec/data_types/data_type_struct.cpp [hunk 6]
```diff
@@ -91,6 +80,94 @@ std::string DataTypeStruct::do_get_name() const {
...
 
+Status DataTypeStruct::from_string(ReadBuffer& rb, IColumn* column) const {
+    DCHECK(!rb.eof());
+    auto* struct_column = assert_cast<ColumnStruct*>(column);
+
+    if (*rb.position() != '{') {
+        return Status::InvalidArgument("Struct does not start with '{' character, found '{}'",
+                                       *rb.position());
+    }
+    if (*(rb.end() - 1) != '}' && rb.count() > 1) {
```
Comment (cambyzju) [inline be/src/vec/data_types/data_type_struct.cpp:91]:
```suggestion
    if (rb.count() < 2 || *(rb.end() - 1) != '}' && ) {
```

File: be/src/vec/data_types/data_type_struct.cpp (hunk 3 above)
Comment (carlvinhust2012) [reply be/src/vec/data_types/data_type_struct.cpp]: done

File: be/src/vec/data_types/data_type_struct.cpp (hunk 6 above)
Comment (carlvinhust2012) [reply be/src/vec/data_types/data_type_struct.cpp]: done

**Review by cambyzju — APPROVED** [review]
LGTM

**Pull request merged**